src/Plutarch/Extra/Precompile.hs

```diff
@@ -0,0 +1,240 @@
+applyCompiledTerm2 (CompiledTerm sf) (CompiledTerm sa) =
+    CompiledTerm $ applyDScript sf (mustEvalD sa)
+
+{- | Apply a 'CompiledTerm' to a 'CompiledTerm'.
+
+ Does NOT evaluate the argument before applying. Using this seems to save very
+ little overhead, not worth it for efficiency. Only use it to make argument
+ evaluation count for benchmarking.
+-}
+applyCompiledTerm2' ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    CompiledTerm a ->
+    CompiledTerm b
+applyCompiledTerm2' (CompiledTerm sf) (CompiledTerm sa) =
+    CompiledTerm $ applyDScript sf sa
+
+-- | Alias for 'applyCompiledTerm'.
+(##) ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    (forall (s :: S). Term s a) ->
+    CompiledTerm b
+(##) = applyCompiledTerm
+
+infixl 8 ##
+
+-- | Alias for 'applyCompiledTerm\''.
+(##~) ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    (forall (s :: S). Term s a) ->
+    CompiledTerm b
+(##~) = applyCompiledTerm'
+
+infixl 8 ##~
+
+-- | Alias for 'applyCompiledTerm2'.
+(###) ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    CompiledTerm a ->
+    CompiledTerm b
+(###) = applyCompiledTerm2
+
+infixl 7 ###
+
+-- | Alias for 'applyCompiledTerm2\''.
+(###~) ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    CompiledTerm a ->
+    CompiledTerm b
+(###~) = applyCompiledTerm2'
+
+infixl 7 ###~
+
+--  Copied and adapted the stuff below from 'Plutarch.Lift'. Including
+--  'LiftError', because the data constructors aren't exported there. Also
+--  added trace messages in the exceptions.
+
+-- | Error during script evaluation.
+data LiftError
+    = LiftError_EvalError EvalError [Text]
+    | LiftError_KnownTypeError KnownTypeError
+    | LiftError_FromRepr
+    deriving stock (Eq)
+
+{- | Convert a 'CompiledTerm' to the associated Haskell value. Fail otherwise.
+
+ This will fully evaluate the compiled term, and convert the resulting value.
+-}
+pliftCompiled' ::
+    forall p. PUnsafeLiftDecl p => CompiledTerm p -> Either LiftError (PLifted p)
+pliftCompiled' prog = case finalEvalDScript (toDScript prog) of
+    (Right (unScript -> UplcType.Program _ _ term), _, _) ->
+        case readKnownConstant term of
+            Right r -> case pconstantFromRepr r of
+                Just h -> Right h
+                Nothing -> Left LiftError_FromRepr
+            Left e -> Left $ LiftError_KnownTypeError e
+    (Left e, _, logs) -> Left $ LiftError_EvalError e logs
+
+-- | Like `pliftCompiled'` but throws on failure.
+pliftCompiled ::
+    forall p.
+    (HasCallStack, PLift p) =>
+    CompiledTerm p ->
+    PLifted p
+pliftCompiled prog = case pliftCompiled' prog of
+    Right x -> x
+    Left LiftError_FromRepr ->
+        error "plift failed: pconstantFromRepr returned 'Nothing'"
+    Left (LiftError_KnownTypeError e) ->
+        let unliftErrMaybe = e ^? _UnliftingErrorE
```

Comment:
Do we really need lens for this?
```hs
    Left (LiftError_KnownTypeError e) ->
       case e of
         KnownTypeUnliftingError unliftErr ->
           error $ "plift failed: incorrect type: " <> show unliftErr
         KnownTypeEvaluationFailure ->
           error $ "plift failed: absurd evaluation failure"
```

Reply:
Yeah lens is not needed. My original intent was to stay as close as possible to the original code from Plutarch to have an easier time porting over changes, but the function is not really complex enough that it matters.